src/libPMacc/include/nvidia/atomic.hpp

```diff
@@ -51,22 +52,28 @@ namespace nvidia
  * nvidia parallel forall example
  * http://devblogs.nvidia.com/parallelforall/cuda-pro-tip-optimized-filtering-warp-aggregated-atomics/
  */
+template<typename T>
 DINLINE
-int atomicAllInc(int *ptr)
+T atomicAllInc(T *ptr)
 {
 #if (__CUDA_ARCH__ >= 300)
-    const int mask = __ballot(1);
-    /* select the leader */
-    const int leader = __ffs(mask) - 1;
-    int restult;
-    const int lanId = getLaneId();
-    if (lanId == leader)
-        restult = atomicAdd(ptr, __popc(mask));
-    restult = warpBroadcast(restult, leader);
-    /* each thread computes its own value */
-    return restult + __popc(mask & ((1 << lanId) - 1));
+    // warpBroadcast is only defined for int (unsigned int is also ok through conversion)
+    if(boost::is_same<T, int>::value || boost::is_same<T, unsigned>::value)
```

Comment:
brr, can't you rather write a specialization of the function for those types ? ;)